src/query_plan/generate.rs

```diff
@@ -0,0 +1,321 @@
+/// But when a "good" plan is generated early, it can save a lot of computing work.
+///
+/// And the 2nd "trick" of this method is that it starts by generating the plans
+/// that correspond to picking choices in `to_add` at the same indexes,
+/// and this because this often actually generate good plans.
+/// The reason is that the order of choices for each element of `to_add` is not necessarily random,
+/// because the algorithm generating paths is not random either.
+/// In other words, elements at similar indexes have some good chance
+/// to correspond to similar choices, and so will tend to correspond to good plans.
+///
+/// Parameters:
+/// * `initial`: the initial partial plan to use.
+/// * `to_add`: a list of the remaining "elements" to add to `initial`.
+///   Each element of `to_add` correspond to multiple choice
+///   we can use to plan that particular element.
+///   The `Option`s in side `type Choices<T> = Vec<Option<T>>` are for internal use
+///   and should all be `Some` when calling this function.
+/// * `add_function`: how to obtain a new plan by taking some plan and adding a new element to it.
+/// * `cost_function`: how to compute the cost of a plan.
+/// * `on_plan`: an optional method called on every _complete_ plan generated by this method,
+///   with both the cost of that plan and the best cost we have generated thus far
+///   (if that's not the first plan generated).
+///   This mostly exists to allow some debugging.
+pub fn generate_all_plans_and_find_best<Plan, Element>(
+    initial: Plan,
+    to_add: Vec<Choices<Element>>,
+    mut add_function: impl FnMut(&Plan, Element) -> Plan,
+    mut cost_function: impl FnMut(&Plan) -> QueryPlanCost,
+    mut on_plan: impl FnMut(&Plan, QueryPlanCost, Option<QueryPlanCost>),
+) -> (Plan, QueryPlanCost)
+where
+    Element: Clone,
+    // Uncomment to use `dbg!()`
+    // Element: std::fmt::Debug,
+{
+    if to_add.is_empty() {
+        let cost = cost_function(&initial);
+        return (initial, cost);
+    }
+    // Note: we save ourselves the computation of the cost of `initial`
+    // (we pass no `partial_cost` in this initialization).
+    // That's because `partial_cost` is about exiting early when we've found at least one full plan
+    // and the cost of that plan is already smaller than the cost of a partial computation.
+    // But any plan is going be built from `initial` with at least one 'choice' added to it,
+    // all plans are guaranteed to be more costly than `initial` anyway.
+    // Note that save for `initial`,
+    // we always compute `partialCost` as the pros of exiting some branches early are large enough
+    // that it outweight computing some costs unecessarily from time to time.
+    let mut stack = VecDeque::new();
+    stack.push_back(Partial {
+        partial_plan: initial,
+        partial_cost: None,
+        remaining: to_add.into_iter(),
+        is_root: true,
+        index: Some(0),
+    });
+
+    let mut min = None;
```

Comment:
wondering whether using `Option` here is better than just initializing it to max i64?

Reply:
guess we could potentially generate with max i64 cost... but that seems unlikely (guess never say never)?

Reply:
`Option` matches the TypeScript version: `let min: { best: P, cost: number } | undefined = undefined;`. Is there a reason to prefer something else?

Reply:
my bad I missed that it is a tuple vs just cost